website/blog/2022-01-31-avoid-data-silos-in-presto-in-meta.md

```diff
@@ -0,0 +1,171 @@
+presto clusters, reducing operational overhead.
+
+### RaptorX Architecture
+Following is the architecture of RaptorX:
+
+![RaptorX Architecture](/img/blog/2022-01-31-avoid-data-silos-in-presto-in-meta/raptorx_architecture.png)
+
+The fundamental difference between Raptor and Raptor X is how the local SSDs on workers are used. In RaptorX, Presto workers use [Alluxio](www.alluxio.io) to cache
+file data locally. It is well understood that access patterns for different table columns could be very different, and columnar file formats
+like ORC and Parquet are commonly used for data files to increase data locality within files. By caching file fragments in small page size on top of columnar files,
+only data that are frequently accessed will be kept close to compute. Presto coordinator tries to schedule compute that processes the same data to the same worker node to
+increase cache effectiveness. RaptorX also implements file footer and metadata caching, and other smart caching strategies that would improve
+the performance further.
+
+For more details about RaptorX architecture, please read the blog [RaptorX: Building a 10X Faster Presto](https://prestodb.io/blog/2021/02/04/raptorx).
+
+### RaptorX vs Raptor performance benchmark
+We ran benchmarks to compare the performance of prototyped RaptorX against Raptor. The benchmark is run on a cluster with ~1000 worker nodes and a single coordinator.
+Raptor and RaptorX are using the same hardware, so the whole dataset fits in RaptorX local SSD cache, thus cache hit rate is close to 100%.
+
+![Benchmark](/img/blog/2022-01-31-avoid-data-silos-in-presto-in-meta/benchmark.png)
+
+As you can see from the benchmark result, P90 latency has an almost 2x improvement for RaptorX compared to Raptor. The difference between average query latency
+and P90 query latency in RaptorX is much smaller compared to Raptor. This is because in Raptor, data is physically bound to the worker node hosting it, thus a
+slow node would inevitably affect query latency. In RaptorX, instead of hard affinity between worker and data, we use soft affinity when scheduling. Soft affinity
+will select two worker nodes as candidates to process a split. If the first choice worker node is up and healthy, that node would be chosen, otherwise a secondary
+node will be chosen. Data can potentially be cached at multiple nodes, and scheduling can optimize for better CPU load balancing for the overall workloads.
+
+### Migration from Raptor to RaptorX
+All previous Raptor use cases in Meta are migrated to RaptorX, which provides better user experience and is easy to scale.
+
+#### A/B Testing Framework
+In the previous section we mentioned that the requirements for the A/B testing framework are: losslessness, flexibility, freshness, interactive latency and high availability.
+Since RaptorX is a caching layer on the original Hive data, losslessness is guaranteed by Hive. It enjoys all the query optimization from the core presto engine, as well as many
```

Comment:
```suggestion
Since RaptorX is a caching layer on the original Hive data, losslessness is guaranteed by Hive. It enjoys all the query optimization from the core Presto engine, as well as many
```